prereise/gather/translate_zones.py

```diff
@@ -0,0 +1,229 @@
+    )
+
+    # pivots dataframe into a matrix where rows=prev_zones and cols=new_zones
+    prev_to_new_matrix = prev_overlapped_by_new.pivot(
+        index=name_prev, columns=name_new, values="pct_of_prev_area"
+    )
+
+    # clean NaNs
+    prev_to_new_matrix = prev_to_new_matrix.rename(columns={np.nan: "none"})
+    prev_to_new_matrix = prev_to_new_matrix.fillna(value=0)
+
+    # check for prev_zones that do not overlap new_zones, then remove
+    isolated_zones = prev_to_new_matrix.loc[prev_to_new_matrix["none"] >= 1]
+    if len(isolated_zones > 0):
+        print(
+            f"\nWARNING: Detected {name_prev} zones that do not overlap with {name_new}: {list(isolated_zones.index)}"
+        )
+        prev_to_new_matrix = prev_to_new_matrix.loc[prev_to_new_matrix["none"] < 1]
+
+    # Alert if sum > 1 (e.g. new zones overlap each other)
+    # NOTE: only checks to three decimal places
+    prev_to_new_matrix["sum"] = prev_to_new_matrix.sum(axis=1)
+    sum_greater_than_one = prev_to_new_matrix.loc[
+        prev_to_new_matrix["sum"].round(3) > 1
+    ]
+    if len(sum_greater_than_one > 0):
+        print(
+            f"WARNING: Detected rows with a sum greater than one. This most likely means that several {name_new} zones overlap each other."
+        )
+        if verbose:
+            print("\nROWS WITH SUM GREATER THAN ONE:")
+        if verbose:
+            print(sum_greater_than_one.round(3))
+
+    # Scale matrix such that all rows add up to 1 by distributing demand mismatch
+    # proportionally between new zones.
+    # if new zones overlap each other, we may end up with > 100% coverage of old zones
+    # if old zones aren't completely covered by new zones, we have < 100% coverage
+    # Note: we could technically simplify this to just scale_factor = 1/(sum - none) but
+    # the extra cols are nice for verbose mode
+    prev_to_new_matrix["sum_without_none"] = (
+        prev_to_new_matrix["sum"] - prev_to_new_matrix["none"]
+    )
+    prev_to_new_matrix["to_distribute"] = 1 - prev_to_new_matrix["sum_without_none"]
+    prev_to_new_matrix["scale_factor"] = (
+        1 + prev_to_new_matrix["to_distribute"] / prev_to_new_matrix["sum_without_none"]
+    )
+
+    if verbose:
+        print("\nMATRIX BEFORE SCALING")
+    if verbose:
+        print(prev_to_new_matrix.round(3))
+
+    scaled_matrix = prev_to_new_matrix.mul(
+        prev_to_new_matrix["scale_factor"], axis="rows"
+    )
+    scaled_matrix = scaled_matrix.drop(
+        columns=["none", "sum", "sum_without_none", "to_distribute", "scale_factor"]
+    )
+
+    # Add any rows that got dropped because they do not overlap. Fill with zeros.
+    # NOTE: not sure if this will get slow with larger datasets
+    if len(scaled_matrix) < len(prev_zones):
+        index = scaled_matrix.index.values
+        for zone in list(prev_zones[name_prev]):
+            if not zone in index:
+                scaled_matrix.loc[zone] = 0
```

Comment:
Maybe we can do following here:
```python
scaled_matrix.loc[list(set(prev_zones[name_prev]) - set(scaled_matrix.index))] = 0
```
Similar for adding columns below.

Reply:
I believe assigning with loc only works if you do one row at a time. If one is missing we get:
`"None of [Index(['c'], dtype='object', name='gray_zones')] are in the [index]"`

Using a for loop for now, but let me know if you know a better way!
```python
new_rows = list(set(prev_zones[name_prev]) - set(scaled_matrix.index))
for row in new_rows:
    scaled_matrix.loc[row] = 0
```

Reply:
`pandas.concat` will let you add an arbitrary number of rows at once, but doesn't modify inplace like a `loc` or `append` addition.

Reply:
Yeah, you are right. Adding multiple rows in-place in one shot is not supported. Other than using a loop, using `append`/`concat` should work as well, not in-place operation though:
```python
scaled_matrix  = pd.concat([scaled_matrix, pd.DataFrame(0, index=new_rows, columns=scaled_matrix.columns)])
```